kms/capi/capi.go

```diff
@@ -0,0 +1,646 @@
+	}
+
+	kh, err := nCryptCreatePersistedKey(k.providerHandle, k.containerName, alg, AT_KEYEXCHANGE, 0)
+	if err != nil {
+		return nil, errors.Wrap(err, "unable to create persisted key")
+	}
+
+	defer nCryptFreeObject(kh)
+
+	if alg == "RSA" {
+		err = nCryptSetProperty(kh, NCRYPT_LENGTH_PROPERTY, uint32(req.Bits), 0)
+
+		if err != nil {
+			return nil, fmt.Errorf("unable to set key NCRYPT_LENGTH_PROPERTY: %w", err)
+		}
+	}
+
+	// users can store the key as a machine key by passing in storelocation = machine
+	// 'machine' is the only valid location, otherwise the key is stored as a 'user' key
+	storeLocation := u.Get(StoreLocationArg)
+
+	if storeLocation == "machine" {
+		err = nCryptSetProperty(kh, NCRYPT_KEY_TYPE_PROPERTY, NCRYPT_MACHINE_KEY_FLAG, 0)
+
+		if err != nil {
+			return nil, fmt.Errorf("unable to set key NCRYPT_KEY_TYPE_PROPERTY: %w", err)
+		}
+	} else if storeLocation != "" && storeLocation != "user" {
+		return nil, fmt.Errorf("invalid storeLocation %v", storeLocation)
+	}
+
+	// if supplied, set the smart card pin
+	scPIN := u.Pin()
+
+	//failover to pin set in kms instantiation
+	if scPIN == "" {
+		scPIN = k.pin
+	}
+
+	if scPIN != "" {
+		err = nCryptSetProperty(kh, NCRYPT_PIN_PROPERTY, scPIN, 0)
+
+		if err != nil {
+			return nil, fmt.Errorf("unable to set key NCRYPT_PIN_PROPERTY: %w", err)
+		}
+	}
+
+	err = nCryptFinalizeKey(kh, 0)
+	if err != nil {
+		return nil, fmt.Errorf("unable to finalize key: %w", err)
+	}
+
+	uc, err := nCryptGetPropertyStr(kh, NCRYPT_UNIQUE_NAME_PROPERTY)
+	if err != nil {
+		return nil, fmt.Errorf("unable to retrieve NCRYPT_UNIQUE_NAME_PROPERTY: %w", err)
+	}
+
+	pub, err := getPublicKey(kh)
+	if err != nil {
+		return nil, fmt.Errorf("unable to retrieve public key: %w", err)
+	}
+
+	return &apiv1.CreateKeyResponse{
+		Name:      uc,
+		PublicKey: pub,
+		CreateSignerRequest: apiv1.CreateSignerRequest{
+			SigningKey: uc,
+		},
+	}, nil
+}
+
+// GetPublicKey returns the public key from the file passed in the request name.
+func (k *CAPIKMS) GetPublicKey(req *apiv1.GetPublicKeyRequest) (crypto.PublicKey, error) {
+	var u *uri.URI
+
+	u, err := uri.ParseWithScheme(Scheme, req.Name)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse URI: %w", err)
+	}
+
+	if k.containerName = u.Get(ContainerNameArg); k.containerName == "" {
+		return nil, fmt.Errorf("GetPublicKey %v not specified", ContainerNameArg)
+	}
+
+	kh, err := nCryptOpenKey(k.providerHandle, k.containerName, AT_KEYEXCHANGE, 0)
+	if err != nil {
+		return nil, fmt.Errorf("unable to open key: %w", err)
+	}
+
+	defer nCryptFreeObject(kh)
+
+	return getPublicKey(kh)
+}
+
+func (k *CAPIKMS) LoadCertificate(req *apiv1.LoadCertificateRequest) (*x509.Certificate, error) {
+	u, err := uri.ParseWithScheme(Scheme, req.Name)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse URI: %w", err)
+	}
+
+	var sha1Hash string
+	if sha1Hash = u.Get(HashArg); sha1Hash == "" {
+		return nil, fmt.Errorf("%v is required", HashArg)
+	}
```

Comment:
You mentioned before that you could also look by serial number, right? 

If this is possible, I would probably add both options. If `sha1` is present, use it, if not `serial`, and fail if both are not present. 

If both are present and you retrieve it by sha1, you should fail if the serial does not match.

Reply:
ive added searching for issuer+serial, and subject key id